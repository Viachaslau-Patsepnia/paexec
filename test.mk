CLEANFILES +=		_test.in

# the following section will be fixed/improved in the next version of mk-c
.if defined(MAKEOBJDIR)
OBJDIR_paexec ?=	${MAKEOBJDIR}
OBJDIR_broken_echo ?=	${MAKEOBJDIR}
.else
OBJDIR_paexec ?=	${.CURDIR}/paexec
OBJDIR_broken_echo ?=	${.CURDIR}/tests/broken_echo
.endif

#
.PHONY : test
test : all-paexec all-examples all-tests
	@echo 'running tests...'; \
	export OBJDIR=${.OBJDIR}; \
	export PATH=${OBJDIR_broken_echo}:${OBJDIR_paexec}:$$PATH; \
	if cd ${.CURDIR}/tests && ./test.sh; \
	then echo 'SUCCEEDED'; \
	else echo 'FAILED'; false; \
	fi
