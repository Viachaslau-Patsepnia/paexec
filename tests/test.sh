#!/bin/sh

set -e

runtest (){
    printf '=================================================================\n'
    printf '======= args: %s\n' "$*"
    ../paexec "$@"
}

cut_version (){
    awk '$1 == "paexec" {$2 = "x.y.x"} {print}'
}

cut_help (){
    awk 'NR <= 3'
}

resort (){
    awk '{print $1, NR, $0}' |
    sort -k1,1n -k2,2n |
    awk '{$1 = $2 = ""; print substr($0, 3)}'
}

do_test (){
    runtest -V        | cut_version
    runtest --version | cut_version

    runtest -h 2>&1     | cut_help
    runtest --help 2>&1 | cut_help

    # toupper
    printf 'a\nbb\nccc\ndddd\neeeee\nffffff\n' |
    runtest -l -t ./paexec_notransport -c ../examples/toupper/toupper_cmd \
	-n '1 2 3 4 5 6 7 8 9' | resort

    # toupper
    printf 'a\nbb\nccc\ndddd\neeeee\nffffff\n' |
    runtest -l -p -t ./paexec_notransport \
	-c ../examples/toupper/toupper_cmd -n '+2' |
    resort | gawk '$1 ~ /^[0-9]/ {$2 = "pid"; print; next} {print}'

    # all_substr
    printf 'a\nbb\nccc\ndddd\neeeee\nffffff\n' |
    runtest -l -c ../examples/all_substr/all_substr_cmd \
	-n '1 2 3 4 5 6 7 8 9' | resort

    # all_substr
    printf 'a\nbb\nccc\ndddd\neeeee\nffffff\n' |
    runtest -l -c ../examples/all_substr/all_substr_cmd -n '+9' |
    resort

    # bi-i-i-i-i-i-ig result
    awk '
    BEGIN {
	for (i=0; i < 10; ++i) {
	    print "1234567890-=qwertyuiop[]asdfghjkl;zxcvbnm,./zaqwsxcderfvbgtyhnmjuik,.lo";
	}
    }' | runtest -c ../tests/big_result_cmd -n '+9' |
    uniq -c |
    head -n 100 |
    awk '{$1 = $1; print $0}'
}

for PAEXEC_BUFSIZE in 1 2 3 4 5 6 7 8 9 10 11 12 13 14; do
    printf "PAEXEC_BUFSIZE=%d:\n" $PAEXEC_BUFSIZE
    export PAEXEC_BUFSIZE

    if do_test > _test.res && diff -u test.out _test.res; then
	true
    else
	echo "paexec fails (PAEXEC_BUFSIZE=$PAEXEC_BUFSIZE)" 1>&2
	exit 1
    fi
    printf "done\n\n"
done
