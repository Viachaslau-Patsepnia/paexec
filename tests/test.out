=================================================================
======= args: -V
paexec x.y.x written by Aleksey Cheusov
=================================================================
======= args: --version
paexec x.y.x written by Aleksey Cheusov
=================================================================
======= args: -h
paexec - parallel executor
=================================================================
======= args: --help
paexec - parallel executor
=================================================================
======= args: -l -t ./paexec_notransport -c ../examples/toupper/toupper_cmd -n 1 2 3 4 5 6 7 8 9
0 A
1 BB
2 CCC
3 DDDD
4 EEEEE
5 FFFFFF
=================================================================
======= args: -l -p -t ./paexec_notransport -c ../examples/toupper/toupper_cmd -n +2
0 pid A
1 pid BB
2 pid CCC
3 pid DDDD
4 pid EEEEE
5 pid FFFFFF
=================================================================
======= args: -l -c ../examples/all_substr/all_substr_cmd -n 1 2 3 4 5 6 7 8 9
0 substr[1,1]=a
1 substr[1,1]=b
1 substr[1,2]=bb
1 substr[2,1]=b
2 substr[1,1]=c
2 substr[1,2]=cc
2 substr[1,3]=ccc
2 substr[2,1]=c
2 substr[2,2]=cc
2 substr[3,1]=c
3 substr[1,1]=d
3 substr[1,2]=dd
3 substr[1,3]=ddd
3 substr[1,4]=dddd
3 substr[2,1]=d
3 substr[2,2]=dd
3 substr[2,3]=ddd
3 substr[3,1]=d
3 substr[3,2]=dd
3 substr[4,1]=d
4 substr[1,1]=e
4 substr[1,2]=ee
4 substr[1,3]=eee
4 substr[1,4]=eeee
4 substr[1,5]=eeeee
4 substr[2,1]=e
4 substr[2,2]=ee
4 substr[2,3]=eee
4 substr[2,4]=eeee
4 substr[3,1]=e
4 substr[3,2]=ee
4 substr[3,3]=eee
4 substr[4,1]=e
4 substr[4,2]=ee
4 substr[5,1]=e
5 substr[1,1]=f
5 substr[1,2]=ff
5 substr[1,3]=fff
5 substr[1,4]=ffff
5 substr[1,5]=fffff
5 substr[1,6]=ffffff
5 substr[2,1]=f
5 substr[2,2]=ff
5 substr[2,3]=fff
5 substr[2,4]=ffff
5 substr[2,5]=fffff
5 substr[3,1]=f
5 substr[3,2]=ff
5 substr[3,3]=fff
5 substr[3,4]=ffff
5 substr[4,1]=f
5 substr[4,2]=ff
5 substr[4,3]=fff
5 substr[5,1]=f
5 substr[5,2]=ff
5 substr[6,1]=f
=================================================================
======= args: -l -c ../examples/all_substr/all_substr_cmd -n +9
0 substr[1,1]=a
1 substr[1,1]=b
1 substr[1,2]=bb
1 substr[2,1]=b
2 substr[1,1]=c
2 substr[1,2]=cc
2 substr[1,3]=ccc
2 substr[2,1]=c
2 substr[2,2]=cc
2 substr[3,1]=c
3 substr[1,1]=d
3 substr[1,2]=dd
3 substr[1,3]=ddd
3 substr[1,4]=dddd
3 substr[2,1]=d
3 substr[2,2]=dd
3 substr[2,3]=ddd
3 substr[3,1]=d
3 substr[3,2]=dd
3 substr[4,1]=d
4 substr[1,1]=e
4 substr[1,2]=ee
4 substr[1,3]=eee
4 substr[1,4]=eeee
4 substr[1,5]=eeeee
4 substr[2,1]=e
4 substr[2,2]=ee
4 substr[2,3]=eee
4 substr[2,4]=eeee
4 substr[3,1]=e
4 substr[3,2]=ee
4 substr[3,3]=eee
4 substr[4,1]=e
4 substr[4,2]=ee
4 substr[5,1]=e
5 substr[1,1]=f
5 substr[1,2]=ff
5 substr[1,3]=fff
5 substr[1,4]=ffff
5 substr[1,5]=fffff
5 substr[1,6]=ffffff
5 substr[2,1]=f
5 substr[2,2]=ff
5 substr[2,3]=fff
5 substr[2,4]=ffff
5 substr[2,5]=fffff
5 substr[3,1]=f
5 substr[3,2]=ff
5 substr[3,3]=fff
5 substr[3,4]=ffff
5 substr[4,1]=f
5 substr[4,2]=ff
5 substr[4,3]=fff
5 substr[5,1]=f
5 substr[5,2]=ff
5 substr[6,1]=f
1 =================================================================
1 ======= args: -c ../tests/big_result_cmd -n +9
100000 1234567890-=QWERTYUIOP[]ASDFGHJKL;ZXCVBNM,./ZAQWSXCDERFVBGTYHNMJUIK,.LO
